contracts/src/StakeTable.sol

```diff
@@ -191,18 +217,66 @@ contract StakeTable is AbstractStakeTable {
         override
         returns (uint64, uint64)
     {
-        bytes32 hash = _hashBlsKey(blsVK);
-        nodes[hash].balance += amount;
-        return (0, 0);
+        bytes32 key = _hashBlsKey(blsVK);
+        Node memory node = nodes[key];
+
+        // The deposit must come from the node's registered account.
+        if (node.account != msg.sender) {
+            revert CallerIsNotTheNodeOwner();
+        }
+
+        // A node cannot deposit more tokens while it waiting to register,
+        if (currentEpoch() <= node.registerEpoch) {
+            revert PrematureDeposit();
+        }
+
+        // A node cannot deposit more tokens if an exit request is in progress.
+        if (node.exitEpoch != 0) {
+            revert ExitRequestInProgress();
+        }
+
+        node.balance += amount;
+        nodes[key] = node;
+        SafeTransferLib.safeTransferFrom(ERC20(tokenAddress), msg.sender, address(this), amount);
+
+        emit Deposit(_hashBlsKey(blsVK), uint256(amount));
+
+        uint64 effectiveEpoch = currentEpoch() + 1;
+
+        return (node.balance, effectiveEpoch);
     }
 
     /// @notice Request to exit from the stake table, not immediately withdrawable!
     ///
     /// @param blsVK The BLS verification key to exit
     /// @return success status
     function requestExit(BN254.G2Point memory blsVK) external override returns (bool) {
-        bytes32 hash = _hashBlsKey(blsVK);
-        nodes[hash].exitEpoch = 0;
+        bytes32 key = _hashBlsKey(blsVK);
+        Node memory node = nodes[key];
+
+        // The exit request must come from the node's withdrawal account.
+        if (node.account != msg.sender) {
+            revert CallerIsNotTheNodeOwner();
+        }
+
+        // Cannot request to exit if an exit request is already in progress.
+        if (node.exitEpoch != 0) {
+            revert ExitRequestInProgress();
+        }
+
+        // Cannot exit before becoming an active participant. Activation happens one epoch after the
+        // node's registration epoch, due to the consensus-imposed activation waiting period.
+        if (currentEpoch() < node.registerEpoch + 1) {
+            revert PrematureExit();
+        }
+
+        // Prepare the node to exit.
+        node.exitEpoch = this.nextExitEpoch();
+
+        nodes[key] = node;
```

Comment:
maybe just `nodes[key].exitEpoch = this.nextExitEpoch()` instead of copying the whole struct?

Reply:
Similarly to the [thread](https://github.com/EspressoSystems/espresso-sequencer/pull/791#discussion_r1415815687) above, some test fails when doing:

```solidity
[nix-shell:~/repositories/espressosys/espresso-sequencer]$ git diff
diff --git a/contracts/src/StakeTable.sol b/contracts/src/StakeTable.sol
index 54f2cd4..416f2e4 100644
--- a/contracts/src/StakeTable.sol
+++ b/contracts/src/StakeTable.sol
@@ -271,9 +271,7 @@ contract StakeTable is AbstractStakeTable {
         }
 
         // Prepare the node to exit.
-        node.exitEpoch = this.nextExitEpoch();
-
-        nodes[key] = node;
+        nodes[key].exitEpoch = this.nextExitEpoch();
 
         emit Exit(key, node.exitEpoch);
```

```
Failing tests:
Encountered 1 failing test in contracts/test/StakeTable.t.sol:StakeTable_Test
[FAIL. Reason: log != expected log] test_RequestExit_Succeeds() (gas: 453974)
```

Reply:
Also a bug on my side. Fixed in 8e9c15a.